### ISSUE TITLE
Zenith does not start

### ISSUE THREAD

**Issue by iicurtis** [bvaisvil/zenith#20]
## What happens

Zenith launches to a blank screen and appears to hang. No information is shown. No response to ^C or ^D. 

I tried using `--disable-history`, but it didn't change anything.

## System Info

Version: 0.7.7
Distro: Arch Linux 5.5.13-zen2-1-zen
Terminals: Kitty, Alacritty, Konsole

```
$ hwinfo --short
cpu:
                       AMD Ryzen 7 1700X Eight-Core Processor, 2770 MHz
graphics card:
                       ATI RX Vega64
storage:
                       Samsung Electronics NVMe SSD Controller SM961/PM961
                       ASRock In Fatal1ty X370 Professional Gaming
                       AMD FCH SATA Controller [AHCI mode]
network:
  enp37s0              Intel I211 Gigabit Network Connection
  wlp36s0              Intel Dual Band Wireless-AC 7265
```

**Comment by bvaisvil** [comment]
Thanks for reporting. Is it possible to give the latest version a try (v.8.0.0) ?

Did you install from AUR?

**Comment by bvaisvil** [comment]
Yeah, forgot to push the version bump. I just rectified that. I don't think that'll make a difference in this issue though.

Do you have a ~/.zenith dir? I'm curious how large it is. There was a previous issue with it generating databases that were very large and it would spend a large amount of time processing them upon startup.

Though, that shouldn't be the case if you run with `--disable-history`. Hmm.

I'll give this a try on my arch VM and see what I find.

**Comment by iicurtis** [comment]
The directory was created but is empty except for a .zenith.lock file.

**Comment by iicurtis** [comment]
![image](https://user-images.githubusercontent.com/29741521/78301557-7d3abe00-74f6-11ea-8b43-11e82823685f.png)

After ~15 minutes it opened! But now its not refreshing.

**Comment by bvaisvil** [comment]
When you're running zenith in this state, how much memory is it using? Is it using a lot of CPU?

**Comment by iicurtis** [comment]
It doesn't seem to use much of either. 30MB of memory, nearly no CPU usage.

![image](https://user-images.githubusercontent.com/29741521/78457657-1a1d6880-7669-11ea-9823-3471d5bc2900.png)

**Comment by bvaisvil** [comment]
This is definitely not behavior I've seen before. I've got zenith running on an arch VM. It runs as expected,  but I don't have a window manager installed on it. 

About the only thing I can think of doing is making a version that logs what it is doing and try to trace down the issue from there, if that's something you'd like to try...

**Comment by iicurtis** [comment]
Sure, I can start debugging some things. Was there something specific you had in mind? I see there is a tui-log, but I'm not sure if it'd be helpful here.

**Comment by bvaisvil** [comment]
I've added some logging to the master branch. 
After zenith is built setting the environmental variable `RUST_LOG` will turn on logging to stderr; this will overwrite the screen, but we can pipe stderr to a file with `2>log.txt`. So for example:
`RUST_LOG=debug ./zenith 2>log.txt`

If you could run this until you see zenith actually show the interface and perform a few updates, that could be very helpful in tracking down the issue.

Thank you very much for you assistance!

**Comment by iicurtis** [comment]
Here's the log with debug:
```
log-debug.txt
   Compiling zenith v0.8.2
    Finished dev [unoptimized + debuginfo] target(s) in 6.08s
     Running `target/debug/zenith`
[2020-04-12T16:14:24Z INFO  zenith] Starting zenith 0.8.2
[2020-04-12T16:14:24Z DEBUG zenith] Starting with Arguments: rate: 2000, cpu: 10, net: 10, disk: 10, process: 8, disable_history: false, db_path: /home/atraii/.zenith
[2020-04-12T16:14:24Z DEBUG zenith] Creating Lock
[2020-04-12T16:14:24Z DEBUG zenith] Initializing Terminal
[2020-04-12T16:20:25Z DEBUG zenith::render] Starting Main Loop.
[2020-04-12T16:20:25Z DEBUG tui::backend::termion] 5095 instructions outputed.
[2020-04-12T16:20:25Z DEBUG zenith::render] Event Tick
[2020-04-12T16:20:25Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Updating CPU
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Adding 2 to cpu_usage_histogram chart.
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Adding 33 to mem_utilization chart.
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Adding 189999 to net_in chart.
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Adding 58422 to net_out chart.
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Updating Process List
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Sorting Process Table
[2020-04-12T16:21:55Z DEBUG zenith::metrics] Updating Frequency
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Updating Disks
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Adding 0 to disk_read chart.
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Adding 0 to disk_write chart.
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Updating Platform
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Updating Network Interfaces
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Updating Batteries.
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Updated Metrics for 387 processes.
[2020-04-12T16:21:56Z DEBUG tui::backend::termion] 4078 instructions outputed.
[2020-04-12T16:21:56Z DEBUG zenith::render] Event Tick
[2020-04-12T16:21:56Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:23:26Z DEBUG zenith::metrics] Updating CPU
[2020-04-12T16:23:26Z DEBUG zenith::metrics] Adding 1 to cpu_usage_histogram chart.
...
```
with the release version:
```
log-release.txt
    Finished release [optimized] target(s) in 0.09s
     Running `target/release/zenith`
[2020-04-12T16:43:50Z INFO  zenith] Starting zenith 0.8.2
[2020-04-12T16:43:50Z DEBUG zenith] Starting with Arguments: rate: 2000, cpu: 10, net: 10, disk: 10, process: 8, disable_history: false, db_path: /home/atraii/.zenith
[2020-04-12T16:43:50Z DEBUG zenith] Creating Lock
[2020-04-12T16:43:50Z DEBUG zenith] Initializing Terminal
[2020-04-12T16:49:51Z DEBUG zenith::render] Starting Main Loop.
[2020-04-12T16:49:51Z DEBUG tui::backend::termion] 5095 instructions outputed.
[2020-04-12T16:49:51Z DEBUG zenith::render] Event Tick
[2020-04-12T16:49:51Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating CPU
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Adding 5 to cpu_usage_histogram chart.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Adding 39 to mem_utilization chart.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Adding 7032653 to net_in chart.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Adding 194923 to net_out chart.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Process List
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Sorting Process Table
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Frequency
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Disks
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Adding 0 to disk_read chart.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Adding 0 to disk_write chart.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Platform
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Network Interfaces
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Batteries.
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updated Metrics for 393 processes.
[2020-04-12T16:51:22Z DEBUG tui::backend::termion] 4785 instructions outputed.
[2020-04-12T16:51:22Z DEBUG zenith::render] Event Tick
[2020-04-12T16:51:22Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating CPU
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Adding 3 to cpu_usage_histogram chart.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Adding 39 to mem_utilization chart.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Adding 37900881 to net_in chart.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Adding 642872 to net_out chart.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Process List
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Sorting Process Table
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Frequency
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Disks
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Adding 235520 to disk_read chart.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Adding 2316288 to disk_write chart.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Platform
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Network Interfaces
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Batteries.
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updated Metrics for 393 processes.
[2020-04-12T16:52:52Z DEBUG tui::backend::termion] 3832 instructions outputed.
[2020-04-12T16:52:52Z DEBUG zenith::render] Event Tick
[2020-04-12T16:52:52Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating CPU
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Adding 6 to cpu_usage_histogram chart.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Adding 42 to mem_utilization chart.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Adding 28853948 to net_in chart.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Adding 535708 to net_out chart.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Process List
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Sorting Process Table
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Frequency
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Disks
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Adding 4167680 to disk_read chart.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Adding 10178560 to disk_write chart.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Platform
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Network Interfaces
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Batteries.
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updated Metrics for 393 processes.
[2020-04-12T16:54:22Z DEBUG tui::backend::termion] 3643 instructions outputed.
[2020-04-12T16:54:22Z DEBUG zenith::render] Event Tick
[2020-04-12T16:54:22Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Updating CPU
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Adding 5 to cpu_usage_histogram chart.
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Adding 38 to mem_utilization chart.
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Adding 30415424 to net_in chart.
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Adding 485687 to net_out chart.
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Updating Process List
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Sorting Process Table
[2020-04-12T16:55:52Z DEBUG zenith::metrics] Updating Frequency
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Updating Disks
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Adding 4927488 to disk_read chart.
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Adding 2881536 to disk_write chart.
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Updating Platform
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Updating Network Interfaces
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Updating Batteries.
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Updated Metrics for 395 processes.
[2020-04-12T16:55:53Z DEBUG tui::backend::termion] 3433 instructions outputed.
[2020-04-12T16:55:53Z DEBUG zenith::render] Event Tick
[2020-04-12T16:55:53Z DEBUG zenith::metrics] Updating Metrics
[2020-04-12T16:57:23Z DEBUG zenith::metrics] Updating CPU
...
```

**Comment by bvaisvil** [comment]
Thanks for this. It does really narrow down the surface area though nothing specific jumps out at me yet.

**Comment by bvaisvil** [comment]
Best guess I've got at this point is that `self.system.refresh_all();` seems to be taking far too long. I just noticed that in the screenshot you posted that zenith is in state 'D', I'm starting to wonder if this is related to https://rachelbythebay.com/w/2014/10/27/ps/.  I think it's waiting for something in the `procfs`, which shoudn't be slow. 

I've added some more logging including in one of my dependency crates. My hope is that if you were to run this again with the additional logging we would discover which part of the procfs it is waiting for.

**Comment by iicurtis** [comment]
The short version is that it appears to be waiting on disks. After reading that link, I did notice there was a problem. One of my disks was set to `noauto,x-systemd.automount`, but the file system was corrupted. I'm not sure what was triggering it, but the 90s seems to be related to the systemd automount timeout. I've disabled that corrupted drive in my fstab and I'll be rebooting here shortly to see if it fixed it.

Do you know why this trips zenith up but not htop, ytop, or bottom?

[log-release.txt](https://github.com/bvaisvil/zenith/files/4487786/log-release.txt)

**Comment by bvaisvil** [comment]
Thank you for the log! I think we're getting close. Good find on that file system.

> Do you know why this trips zenith up but not htop, ytop, or bottom?

This is a great question. I'd have to look at their code, but my guess is that zenith is reading part of `/proc` they must not be or perhaps they read in another thread from their render. Hmm.

**Comment by iicurtis** [comment]
Rebooting after removing the failing systemd-automount entry in my `/etc/fstab` worked. Zenith now works as expected. Do you want me to open a new ticket for handling a slow / hanging procfs?

**Comment by bvaisvil** [comment]
Great to hear that things are working as expected now!

Would be great if you did open a new ticket (referencing this one)!

Thanks for the help in tracking this one down!
